format-common/src/main/java/io/cdap/plugin/format/SchemaDetector.java

```diff
@@ -0,0 +1,160 @@
+   * @throws IOException
+   */
+  @Nullable
+  public Schema detectSchema(String path, FormatContext formatContext,
+                             Map<String, String> fileSystemProperties) throws IOException {
+    if (path == null) {
+      return null;
+    }
+
+    Job job = JobUtils.createInstance();
+    Configuration configuration = job.getConfiguration();
+    for (Map.Entry<String, String> entry : fileSystemProperties.entrySet()) {
+      configuration.set(entry.getKey(), entry.getValue());
+    }
+    for (Map.Entry<String, String> entry : inputFormat.getInputFormatConfiguration().entrySet()) {
+      configuration.set(entry.getKey(), entry.getValue());
+    }
+
+    // hack: the file path should be passed down to the format, which can decide whether to skip it or not
+    String regexPathFilter = null;
+    if ("avro".equalsIgnoreCase(formatName)) {
+      regexPathFilter = ".+\\.avro$";
+    } else if ("parquet".equalsIgnoreCase(formatName)) {
+      regexPathFilter = ".+\\.parquet$";
+    }
+
+    Path fsPath = new Path(path);
+    // need this to load the extra class loader to avoid ClassNotFoundException for the file system
+    FileSystem fs = JobUtils.applyWithExtraClassLoader(job, getClass().getClassLoader(),
+                                                       f -> FileSystem.get(fsPath.toUri(), configuration));
+
+    if (!fs.exists(fsPath)) {
+      throw new IOException("Input path not found");
+    }
+
+    FileStatus fileStatus = fs.getFileStatus(fsPath);
+    if (fileStatus.isFile()) {
+      return inputFormat.detectSchema(formatContext, new FileStreamSupplier(fs, fsPath, fileStatus.getLen()));
+    }
+
+    FileStatus[] files = fs.listStatus(fsPath);
+
+    if (files == null) {
+      throw new IllegalArgumentException("Cannot read files from provided path " + path);
+    }
+
+    if (files.length == 0) {
+      throw new IllegalArgumentException("Provided directory '" + path + "' is empty");
+    }
+
+    Pattern pattern = null;
+    if (!Strings.isNullOrEmpty(regexPathFilter)) {
+      pattern = Pattern.compile(regexPathFilter);
+    }
+
+    for (FileStatus file : files) {
+      if (pattern != null) {
+        Matcher matcher = pattern.matcher(file.getPath().toString());
+        if (!matcher.find()) {
+          continue;
+        }
+      }
+
+      return inputFormat.detectSchema(formatContext, new FileStreamSupplier(fs, fsPath, file.getLen()));
+    }
+
+    throw new IllegalArgumentException(String.format("No file inside \"%s\" matched regex \"%s\"!", path,
+                                                     regexPathFilter));
+  }
+
+  /**
+   * Supplies an InputStream of data to detect schema from.
+   *
+   * It implements Function as a hack to pass the file length down to the Parquet format, which needs it to seek to
+   * the file footer.
+   */
+  private static class FileStreamSupplier implements Supplier<InputStream>, Function<Void, Long> {
```

Comment:
It looks strange to have this class extends `Function` and the input is Void. Is it better to just make this class public, so in format-parquent, you don't need to check if the Supplier is a Function, and call `apply(null)` to get the length?

Reply:
it doesn't work because of classloading. The format isn't able to cast it directly. I will add a comment to clarify this.